23 feature/bulk move users between roles

mariadanieldeepak opened the pull request:
Close #23

mariadanieldeepak committed:
Setup Bulk Move Users page boilerplate

mariadanieldeepak committed:
Implement Bulk Move users by role

mariadanieldeepak committed:
Fix Style CI issue

mariadanieldeepak committed:
Fix Style CI issues

mariadanieldeepak committed:
Fix Style CI issues

mariadanieldeepak committed:
Remove additional line break

mariadanieldeepak committed:
Include line break before return statement

mariadanieldeepak committed:
Remove space after empty keyword

mariadanieldeepak committed:
Include line break before continue statement

sudar commented on assets/js/bulk-move.js:
```diff
@@ -8,10 +8,47 @@
+            return confirm(BULK_MOVE.msg.move_warning);
+        }
+
+        that = this;
```
Let's have this assignment as part of the initial declaration of the variable itself.

mariadanieldeepak replied:
> Let's have this assignment as part of the initial declaration of the variable itself.

Fixed

sudar commented on assets/js/bulk-move.js:
```diff
@@ -8,10 +8,47 @@
+
+        that = this;
+        jQuery.each(BULK_MOVE.validators[currentButton], function (index, validator) {
+            valid &= ! BULK_MOVE[validator](that);
```
Let's use a simple assignment here.

mariadanieldeepak replied:
> Let's use a simple assignment here.

Fixed.

sudar commented on assets/js/bulk-move.js:
```diff
@@ -136,4 +173,6 @@ jQuery(document).ready(function () {
 	jQuery( '#smbm_mbct_selected_term, #smbm_mbct_mapped_term' ).change( function() {
 		jQuery( '.bm_ct_filters, .bm_ct_submit' ).show();
 	});
+
```
Let's not have continuous blank lines.

mariadanieldeepak replied:
> Let's not have continuous blank lines.

 Fixed.

sudar commented on include/BulkMove.php:
```diff
@@ -267,4 +279,12 @@ private function get_post_admin_page() {
 		return $posts_page;
 	}
+
+	private function get_user_admin_page() {
```
Let's add PHPDoc for all functions that are not inherited.

mariadanieldeepak replied:
> Let's add PHPDoc for all functions that are not inherited.

Fixed.

sudar commented on include/Metabox/Base.php:
```diff
@@ -301,4 +301,41 @@ protected function render_result( $posts_moved ) {
 		);
 	}
+
+	protected function render_roles_dropdown( $name = '', $hide_no_role_option = false ) {
```
Let's create two new abstract classes called `PostMetabox` and `UserMetabox` which will inherit `Base` Metabox class.

All common functions that are specific to Post and Users can be moved to these classes. Let's also remove the `BM_Util_Helper` class.

mariadanieldeepak replied:
> Let's create two new abstract classes called PostMetabox and UserMetabox which will inherit Base Metabox class.

Classes created. I've named them as `PostBase` and `UserBase` as we have the name folder, under which these files are present, as `Metabox`

sudar commented on include/Metabox/Base.php:
```diff
@@ -301,4 +301,41 @@ protected function render_result( $posts_moved ) {
+
+		$bulk_move = bulk_move();
+
+		$users_by_roles = $bulk_move->helper->get_users_count_by_roles( $users, $roles );
```
This function will be part of `UserMetabox` and let's remove the dependency to `bulk_move()`.

mariadanieldeepak replied:
> This function will be part of UserMetabox and let's remove the dependency to bulk_move().

Fixed.

sudar commented on include/Metabox/Base.php:
```diff
@@ -301,4 +301,41 @@ protected function render_result( $posts_moved ) {
+		<select id="<?php echo $field_name; ?>"
+				name="<?php echo $field_name; ?>">
+			<?php if ( ! $hide_no_role_option ) : ?>
+				<option value="norole">No role <?php echo '(' . $users_by_roles['norole'] . ')' ?></option>
```
This string should be made available for translation.

mariadanieldeepak replied:
> This string should be made available for translation.

Fixed. As part of this change, you may find code formatting generating diff noise and that was intentional to make the code readable.

We should also be setting a standard for inline HTML's. Either we should stick with Space or tabs for inline HTML. Note: PHPStorm by default includes spaces for inline HTML elements.

sudar commented on include/Metabox/Base.php:
```diff
@@ -301,4 +301,41 @@ protected function render_result( $posts_moved ) {
+
+		?>
+		<select id="<?php echo $field_name; ?>"
+				name="<?php echo $field_name; ?>">
```
Do output sanitization on variables before printing them inside markup.

mariadanieldeepak replied:
> Do output sanitization on variables before printing them inside markup.

Since these values are not from DB, I'm using `sanitize_html_class()`. The reason for using `sanitize_html_class()`is that the values are used in HTML element attributes.

sudar commented on include/Metabox/Users/Role.php:
```diff
@@ -0,0 +1,85 @@
+
+	public function render() {
+		?>
+		<!-- Category Start-->
```
Wrong comment. Looks like a copy/paste error.

mariadanieldeepak replied:
> Wrong comment. Looks like a copy/paste error.

Fixed.

sudar commented on include/Metabox/Users/Role.php:
```diff
@@ -0,0 +1,85 @@
+		return _n( 'Moved %d user between the selected roles', 'Moved %d users between the selected roles.', $posts_moved, 'bulk-move' );
+	}
+
+	public function filter_js_array( $js_array ) {
```
Let's add phpdoc blocks to all functions that are not inherited.

sudar commented on include/Page/Base.php:
```diff
@@ -222,7 +222,10 @@ public function modify_admin_footer_text( $footer_text ) {
 	 * Enqueue JavaScript.
 	 */
 	public function enqueue_script() {
-		wp_enqueue_script( 'bulk-move', $this->get_plugin_dir_url() . 'assets/js/bulk-move.js', array( 'jquery', 'postbox' ), BM_BulkMove::VERSION, true );
```
Let's prevent formatting changes as part of other commits. They just add noise to diffs.

sudar commented on include/Page/User.php:
```diff
@@ -0,0 +1,63 @@
+		$this->slug            = 'bulk-move-users';
+		$this->page_title      = __( 'Bulk Move Users', 'bulk-move' );
+		$this->menu_title      = __( 'Bulk Move Users', 'bulk-move' );
+		$this->warning_message = __( 'WARNING: Posts moved once cannot be reverted. Use with caution.', 'bulk-move' );
```
This should be users not posts.

sudar commented on include/Page/User.php:
```diff
@@ -0,0 +1,63 @@
+	public function register() {
+		parent::register();
+
+		add_filter( 'bm_plugin_action_links', array( $this, 'append_to_plugin_action_links' ) );
```
I think it makes sense to move filter and function this to base class.

mariadanieldeepak replied:
> I think it makes sense to move filter and function this to base class.

Fixed.

sudar replied:
Is this also fixed? I don't see the changes.

sudar commented on include/Page/User.php:
```diff
@@ -0,0 +1,63 @@
+
+		// The meta field should be in the following form.
+		// $meta_field = 'metaboxhidden_bulk-wp_page_' . $this->page_slug;
+		return 'metaboxhidden_bulk-wp_page_bulk-move-users';
```
Make this function generic and then move it to base class.

mariadanieldeepak replied:
> Make this function generic and then move it to base class.

Fixed.

mariadanieldeepak committed:
Initialize var during declaration

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164365187

mariadanieldeepak committed:
Use plain assignment

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164365327

mariadanieldeepak committed:
Remove additional new line

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164366233

mariadanieldeepak committed:
Add PHP Doc block

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164366345

mariadanieldeepak committed:
Remove Util\Helper class

and introduced PostBase & UserBase classes.
Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164366768

mariadanieldeepak committed:
Sanitize variables before use.

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r165266328

mariadanieldeepak committed:
Make string available for translation

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164367127

mariadanieldeepak committed:
Fix incorrect comment

Fix review comment: https://github.com/sudar/bulk-move/pull/73\#discussion_r164367332

mariadanieldeepak committed:
Include PHP Doc block

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164367528

mariadanieldeepak committed:
Avoid noise in diffs

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164367817

mariadanieldeepak committed:
Fix incorrect copy

Fix review comment: https://github.com/sudar/bulk-move/pull/73#discussion_r164367977

sudar commented on assets/js/bulk-move.js:
```diff
@@ -8,10 +8,46 @@
 
 /*jslint browser: true, devel: true*/
 /*global BULK_MOVE, jQuery, document, postboxes, pagenow, ajaxurl*/
+BULK_MOVE.validate_same_user_roles = function(that) {
```
The parameter `that` is not used and could be removed.

mariadanieldeepak replied:
> The parameter `that` is not used and could be removed.

Fixed.

sudar commented on assets/js/bulk-move.js:
```diff
@@ -8,10 +8,46 @@
+            that = this,
+            errorKey;
+
+        if (!currentButton in BULK_MOVE.validators) {
```
You are using a negation in the lhs of this in expression. Are you sure you did not want to negate the result of the expression like so `! (currentButton in BULK_MOVE.validators)`?

mariadanieldeepak replied:
> You are using a negation in the lhs of this in expression. Are you sure you did not want to negate the result of the expression like so `! (currentButton in BULK_MOVE.validators)`?

You're right. It worked in this case, because `!` operator reads right-to-left. However, I added parenthesis.

Reference: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence

sudar commented on assets/js/bulk-move.js:
```diff
@@ -8,10 +8,46 @@
+        jQuery.each(BULK_MOVE.validators[currentButton], function (index, validator) {
+            valid = valid && (!BULK_MOVE[validator](that));
+
+            if (!valid) {
```
There is no return statement if !valid is false. Are you sure this is correct? If so, consider adding return; explicitly.

mariadanieldeepak replied:
> There is no return statement if !valid is false. Are you sure this is correct? If so, consider adding return; explicitly.

I've already included that. May be you missed to pull the changes.

sudar commented on assets/js/bulk-move.js:
```diff
@@ -8,10 +8,46 @@
+        if ( valid ) {
+            return confirm(BULK_MOVE.msg.move_warning);
+        } else {
+            e.preventDefault();
```
There is no return statement in this branch, but you do return something in other branches.

mariadanieldeepak replied:
> There is no return statement in this branch, but you do return something in other branches.

Since this is the last statement, I didn't add it explicitly.

mariadanieldeepak committed:
Move filter to Base class

mariadanieldeepak committed:
Move common function to Base class

mariadanieldeepak committed:
Remove unused var

mariadanieldeepak committed:
Fix JS review comments

mariadanieldeepak committed:
Fix Style CI issue
